### PR TITLE
BE-30: Online Presence & User Status

### PR DESCRIPTION
**Pull request by slo248**
# Description

Document: [Google doc](https://docs.google.com/document/d/1HuOPS_oT9qxxRMt-718B4-rm7p8fKu7hqz28vz5gmkY/edit?usp=sharing)

- Online status model
  - `user_id`: user id
  - `type`: `ONLINE, IDLE, DO_NOT_DISTURB, INVISIBLE, OFFLINE`
  - `last_seen`: the time when last update status occurs
  - `status_text`: the text of the custom text

- Add new socket
- Get statuses from a list of user IDs
- Update online status
- Subscription to online status changed of a user (check friends or not)
- Set threshold heartbeat for each time update status to ONLINE => Set OFFLINE when exceed the threshold
- Add middleware in user schema (after add and before delete then affect the user status collection)

***Note:*** 
- @RLukas2 @ntploc21 you should test the previous graphql requests and APIs since I change `app.listen` to `httpserver.listen`, which makes websocket works.

### CODE REVIEW

File: apollo/src/models/onlineStatus.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+type OnlineStatusType = {
+  uid: String;
+  type: OnlineStatus;
+  last_seen: Date;
+};
+
+const schema = new Schema<OnlineStatusType>(
+  {
+    uid: {
```
Comment (ntploc21) [inline apollo/src/models/onlineStatus.ts:16]: I think the `uid` attribute could be changed to `user_id` to match the database design, with type `mongoose.Types.ObjectId` ([Mongoose ObjectId](https://mongoosejs.com/docs/schematypes.html#objectids)), referencing `User` (you can take a look at [BE-33 PR](https://github.com/mobile-apcs-ntploc21/mobile-backend/pull/13/files#diff-a04818e26e6a66cb6e50851ac025fd0714626f1bc32d8f5374449ffd3b157a53R13)).

Comment (slo248) [reply apollo/src/models/onlineStatus.ts]: I fixed it

**Review by ntploc21 — APPROVED** [review]
LGTM!

**Review by ntploc21 — APPROVED** [review]
I can run it on my machine!

**Pull request merged**